test/core/util/port_isolated_runtime_environment.cc

```diff
@@ -28,22 +31,24 @@
 #include "test/core/util/port.h"
 
 #define MIN_PORT 49152
-#define MAX_PORT 65536
+#define MAX_PORT 65535
 
-int get_random_starting_port() {
+static int get_random_port_offset() {
   srand(gpr_now(GPR_CLOCK_REALTIME).tv_nsec);
-  return rand() % (MAX_PORT - MIN_PORT + 1) + MIN_PORT;
+  double rnd = static_cast<double>(rand()) /
+               (static_cast<double>(RAND_MAX) + 1.0);  // values from [0,1)
+  return static_cast<int>(rnd * (MAX_PORT - MIN_PORT + 1));
 }
 
-static int s_allocated_port = get_random_starting_port();
+static int s_initial_offset = get_random_port_offset();
+static gpr_atm s_pick_counter = 0;
 
 int grpc_pick_unused_port_or_die(void) {
-  int allocated_port = s_allocated_port++;
-  if (s_allocated_port == MAX_PORT) {
-    s_allocated_port = MIN_PORT;
-  }
-
-  return allocated_port;
+  int orig_counter_val =
+      static_cast<int>(gpr_atm_full_fetch_add(&s_pick_counter, 1));
+  GPR_ASSERT(orig_counter_val < (MAX_PORT - MIN_PORT + 1));
```

Comment:
This assert is problematic and will probably start causing issues. I don't see the benefit of this either.

Reply:
why is it problematic?   You want to prevent tests from reusing a port that has been already allocated by another test to prevent hard-to-debug issues caused by crosstalk of tests.  With normal port server, picking the port will also fail if there are no ports available.  The "isolated runtime" implementation doesn't support returning used ports back to port server - so asserting once we exhausted all the available ports is the right thing to do.

Realistically, as each test binary runs in a separate container, exhausting the available port won't happen in practice (you would need a single test binary that goes through several thousand ports), but if someone writes an exotic test that uses too many ports in the future, we do want to know we ran out of ports rather then acting like everything is fine and then end up rootcausing flakes to find out we've been reusing ports silently.

Reply:
oh you are right. I was thinking that we may want to allow tests to start reusing ports

Reply:
Thanks for reviewing!